File: src/main/java/org/stt/config/PasswordSetting.java
package org.stt.config;

import javax.crypto.Cipher;
import javax.crypto.SecretKey;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Arrays;

import static java.util.Objects.requireNonNull;

public class PasswordSetting {
    public static final String CIPHER_ALGORITHM = "AES/ECB/PKCS5Padding";
    private static SecretKey secretKey;
    public final byte[] encodedPassword;

    static {
        MessageDigest md;
        try {
            md = MessageDigest.getInstance("SHA-1");
        } catch (NoSuchAlgorithmException e) {
            throw new IllegalStateException(e);
        }
        byte[] key;
        key = Arrays.copyOf(md.digest("ImagineAReallyStrongPasswordHere".getBytes(StandardCharsets.UTF_8)), 16);
        secretKey = new SecretKeySpec(key, "AES");
    }

    private PasswordSetting(byte[] encodedPassword) {
        requireNonNull(encodedPassword);
        this.encodedPassword = Arrays.copyOf(encodedPassword, encodedPassword.length);
        // Test the encrypted password
        getPassword();
    }

    public byte[] getPassword() {
        try {
            Cipher c = Cipher.getInstance(CIPHER_ALGORITHM);
            c.init(Cipher.DECRYPT_MODE, secretKey);
            return c.doFinal(encodedPassword);
        } catch (Exception e) {
            throw new IllegalStateException(e);
        }
    }

    public static PasswordSetting fromPassword(byte[] password) {
        try {
            Cipher c = Cipher.getInstance(CIPHER_ALGORITHM);
            c.init(Cipher.ENCRYPT_MODE, secretKey);
            return new PasswordSetting(c.doFinal(password));
        } catch (Exception e) {
            throw new IllegalStateException(e);
        }
    }

    public static PasswordSetting fromEncryptedPassword(byte[] password) {
        return new PasswordSetting(password);
    }
}

File: src/main/kotlin/org/stt/config/PasswordSetting.kt
package org.stt.config

import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.util.*
import javax.crypto.Cipher
import javax.crypto.SecretKey
import javax.crypto.spec.SecretKeySpec

class PasswordSetting private constructor(encodedPassword: ByteArray) {
    val encodedPassword: ByteArray
        get() = Arrays.copyOf(field, field.size)

    val password: ByteArray
        get() {
            try {
                val c = Cipher.getInstance(CIPHER_ALGORITHM)
                c.init(Cipher.DECRYPT_MODE, secretKey)
                return c.doFinal(encodedPassword)
            } catch (e: Exception) {
                throw IllegalStateException(e)
            }

        }

    init {
        this.encodedPassword = Arrays.copyOf(encodedPassword, encodedPassword.size)
        // Test the encrypted password
        password
    }

    companion object {
        private const val CIPHER_ALGORITHM = "AES/ECB/PKCS5Padding"
        private var secretKey: SecretKey? = null

        init {
            val md: MessageDigest
            try {
                md = MessageDigest.getInstance("SHA-1")
            } catch (e: NoSuchAlgorithmException) {
                throw IllegalStateException(e)
            }

            val key: ByteArray
            key = Arrays.copyOf(md.digest("ImagineAReallyStrongPasswordHere".toByteArray(StandardCharsets.UTF_8)), 16)
            secretKey = SecretKeySpec(key, "AES")
        }

        fun fromPassword(password: ByteArray): PasswordSetting {
            try {
                val c = Cipher.getInstance(CIPHER_ALGORITHM)
                c.init(Cipher.ENCRYPT_MODE, secretKey)
                return PasswordSetting(c.doFinal(password))
            } catch (e: Exception) {
                throw IllegalStateException(e)
            }

        }

        fun fromEncryptedPassword(password: ByteArray): PasswordSetting {
            return PasswordSetting(password)
        }
    }
}
